jax/_src/numpy/polynomials/chebyshev.py

```diff
@@ -0,0 +1,605 @@
+        x = jnp.asarray(x)
+    if isinstance(x, jnp.ndarray) and tensor:
+        c = c.reshape(c.shape + (1,)*x.ndim)
+
+    if len(c) == 1:
+        c0 = c[0]
+        c1 = 0
+    elif len(c) == 2:
+        c0 = c[0]
+        c1 = c[1]
+    else:
+        x2 = 2*x
+        c0 = c[-2]
+        c1 = c[-1]
+        for i in range(3, len(c) + 1):
+            tmp = c0
+            c0 = c[-i] - c1
+            c1 = tmp + c1*x2
+    return c0 + c1*x
+
+@jit
+def chebval2d(x, y, c):
+    args = [jnp.asarray(x), jnp.asarray(y)]
+    shape0 = args[0].shape
+    if not all((a.shape == shape0 for a in args[1:])):
+        raise ValueError('x, y are incompatible')
+    it = iter(args)
+    x0 = next(it)
+
+    c = chebval(x0, c)
+    for xi in it:
+        c = chebval(xi, c, tensor=False)
+    return c
+
+@jit
+def chebgrid2d(x, y, c):
+    args = [x, y]
+    for xi in args:
+        c = chebval(xi, c)
+    return c
+
+@jit
+def chebval3d(x, y, z, c):
+    args = [jnp.asarray(x), jnp.asarray(y), jnp.asarray(z)]
+    shape0 = args[0].shape
+    if not all((a.shape == shape0 for a in args[1:])):
+        raise ValueError('x, y, z are incompatible')
+    it = iter(args)
+    x0 = next(it)
+
+    c = chebval(x0, c)
+    for xi in it:
+        c = chebval(xi, c, tensor=False)
+    return c
+
+@jit
+def chebgrid3d(x, y, z, c):
+    args = [x, y, z]
+    for xi in args:
+        c = chebval(xi, c)
+    return c
+
+@partial(jit, static_argnames=('deg',))
+def chebvander(x, deg):
+    if deg < 0 or deg != int(deg):
+        raise ValueError("deg must be a non-negative integer")
+    deg = int(deg)
+    x = jnp.array(x, copy=False, ndmin=1) + 0.0
+    dims = (deg+1,) + x.shape
+    dtyp = x.dtype
+    v = jnp.empty(dims, dtype=dtyp)
+    v = v.at[0].set(x*0 + 1)
+    if deg > 0:
+        x2 = 2*x
+        v = v.at[1].set(x)
+        for i in range(2, deg+1):
+            v = v.at[i].set(v[i-1]*x2 - v[i-2])
+    return jnp.moveaxis(v, 0, -1)
+
+@partial(jit, static_argnames=('deg_x', 'deg_y'))
+def chebvander2d(x, y, deg_x, deg_y):
+    points = (x, y)
+    deg = (deg_x, deg_y)
+    # convert to the same shape and type
+    points = tuple(jnp.array(tuple(points), copy=False) + 0.0)
+
+    # produce the vandermonde matrix for each dimension, placing the last
+    # axis of each in an independent trailing axis of the output
+    vander_arrays = (
+        chebvander(points[0], deg[0])[(...,) + (slice(None), jnp.newaxis)],
+        chebvander(points[1], deg[1])[(...,) + (jnp.newaxis, slice(None))]
+    )
+
+    # we checked this wasn't empty already, so no `initial` needed
+    v = functools.reduce(operator.mul, vander_arrays)
+
+    return v.reshape(v.shape[:-len(deg)] + (-1,))
+
+@partial(jit, static_argnames=('deg_x', 'deg_y', 'deg_z'))
+def chebvander3d(x, y, z, deg_x, deg_y, deg_z):
+    points = (x, y, z)
+    deg = (deg_x, deg_y, deg_z)
+    # convert to the same shape and type
+    points = tuple(jnp.array(tuple(points), copy=False) + 0.0)
```

Comment:
This looks like a roundabout way of doing `_promote_dtypes_inexact`; you should instead call that directly.